atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceTransactionIntegrationTest.java

```diff
@@ -60,6 +73,28 @@ protected KeyValueService getKeyValueService() {
         return kvsSupplier.get();
     }
 
+    @Test
+    public void canUntangleHighlyConflictingPutUnlessExists() {
+        TransactionTables.createTables(keyValueService);
+        TransactionTables.truncateTables(keyValueService);
+        TransactionService transactionService = WriteBatchingTransactionService.create(
+                SimpleTransactionService.createV2(keyValueService));
+        ExecutorService executorService = Executors.newCachedThreadPool();
+        List<Future<?>> futures = LongStream.range(0, 100)
+                .mapToObj(unused -> executorService.submit(() -> {
+                    try {
+                        transactionService.putUnlessExists(unused, unused + 1);
+                        transactionService.putUnlessExists(unused + 1, unused + 3);
+                        transactionService.putUnlessExists(unused - 1, unused + 5);
+                    } catch (KeyAlreadyExistsException ex) {
+                        // Some are expected
```

Comment:
We could actually collect and verify that we get the expected number of conflicts for each ts pair

Reply:
Turns out this is trickier than we expected because of the way batches throw out exceptions (the keys reflect everything in the batch). As discussed offline, we'll make assertions on the number of exceptions, but not at the individual ts-pair level.

Reply:
Actually we can't even guarantee that 😞 because of retry logic